libraries/AP_Soaring/SoaringController.cpp

```diff
@@ -0,0 +1,407 @@
+        dx -= dx_w;
+        dy -= dy_w;
+
+        if (0) {
+            // Print filter info for debugging
+            int i;
+            for (i = 0; i < 4; i++) {
+                hal.console->printf("%e ", (double)_ekf.P[i][i]);
+            }
+            for (i = 0; i< 4; i++) {
+                hal.console->printf("%e ", (double)_ekf.X[i]);
+            }
+        } else {
+             // write log - save the data.
+            log_tuning.time_us = AP_HAL::micros64();
+            log_tuning.netto_rate = _vario_reading;
+            log_tuning.dx = dx;
+            log_tuning.dy = dy;
+            log_tuning.x0 = _ekf.X[0];
+            log_tuning.x1 = _ekf.X[1];
+            log_tuning.x2 = _ekf.X[2];
+            log_tuning.x3 = _ekf.X[3];
+            log_tuning.lat = current_loc.lat;
+            log_tuning.lng = current_loc.lng;
+            log_tuning.alt = _alt;
+            log_tuning.dx_w = dx_w;
+            log_tuning.dy_w = dy_w;
+            log_tuning.nsamples = _nsamples;
+        }
+        log_data(); 
+        _ekf.update(_vario_reading,dx, dy);       // update the filter
+         
+        _prev_update_location = current_loc;      // save for next time
+        _prev_update_time = AP_HAL::micros64();
+        _new_data = false;
+    }
+}
+
+void SoaringController::update_cruising()
+{
+    if (is_active() && suppress_throttle()) {
+        init_cruising();
+    }
+}
+
+void SoaringController::update_vario()
+{
+    Location current_loc;
+    _ahrs.get_position(current_loc);
+    _alt = (current_loc.alt -_ahrs.get_home().alt) / 100.0f;
+    
+    if (fabsf(_alt - _last_alt) > 0.0001f) { // if no change in altitude then there will be no update of ekf buffer
+        // Both filtered total energy rates and unfiltered are computed for the thermal switching logic and the EKF
+        float aspd;
+        float roll = _ahrs.roll;
+        if (!_ahrs.airspeed_estimate(&aspd)) {
+            aspd = 0.5f * (_aparm.airspeed_min + _aparm.airspeed_max); //
+        }
+        _aspd_filt = ASPD_FILT * aspd + (1 - ASPD_FILT) * _aspd_filt;
+        float total_E = _alt + 0.5 *_aspd_filt * _aspd_filt / GRAVITY_MSS;                                                  // Work out total energy
+        float sinkrate = correct_netto_rate(0.0f, (roll + _last_roll) / 2, _aspd_filt);                                     // Compute still-air sinkrate
+        _vario_reading = (total_E - _last_total_E) / ((AP_HAL::micros64() - _prev_vario_update_time) * 1e-6) + sinkrate;    // Unfiltered netto rate
+        _filtered_vario_reading = TE_FILT * _vario_reading + (1 - TE_FILT) * _filtered_vario_reading;                       // Apply low pass timeconst filter for noise
+        _displayed_vario_reading = TE_FILT_DISPLAYED * _vario_reading + (1 - TE_FILT_DISPLAYED) * _displayed_vario_reading;
+        
+        Vector2f diff = location_diff(_prev_vario_update_location, current_loc); // get distances from previous update
```

Comment:
Did that... Not sure if it made things that much better though, since the intermediate results computed in these lines are used elsewhere in the code.